### PR TITLE
add -UseDefaultCredentials to New-TssSession

### PR DESCRIPTION
**Pull request by peetrike**
implements #24

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
-* None
+* `New-TssSession` - added -UseDefaultCredentials parameter
```
Comment (wsmelton) [inline CHANGELOG.md:16]: You will need to rebase your branch to pick up the previous merged PR.

As well, this entry should go under the `Added` section.

File: src/classes/TssSession.class.ps1 [hunk 2]
```diff
@@ -19,15 +19,18 @@
     }
 
     [boolean]IsValidToken() {
-        if ([string]::IsNullOrEmpty($this.AccessToken)) {
+        if ($this.TokenType -like 'DefaultCredentials') {
+            Write-Host 'Token is using currently logged on user credentials'
```
Comment (wsmelton) [inline src/classes/TssSession.class.ps1:23]: A token is not actually used with winauth. We can just have this state something like:

```
Write-Host 'Windows Integrated Auth in use
```

File: src/functions/New-Session.ps1 [hunk 3]
```diff
@@ -56,14 +57,22 @@
...
+        [switch]
+        $UseDefaultCredentials
     )
 
     begin {
         $newTssParams = $PSBoundParameters
         $invokeParams = @{ }
 
         $outputTssSession = [TssSession]::new()
+        if ($UseDefaultCredentials.IsPresent) {
```
Comment (wsmelton) [inline src/functions/New-Session.ps1:73]: Please use `$newTssParams.ContainsKey('UseDefaultCredentials')`

**Comment by wsmelton** [comment]
Please run the test to ensure this change does not break anything.

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (peetrike) [reply CHANGELOG.md]: done

File: src/functions/New-Session.ps1 (hunk 3 above)
Comment (peetrike) [reply src/functions/New-Session.ps1]: in 7614719

File: src/classes/TssSession.class.ps1 (hunk 2 above)
Comment (peetrike) [reply src/classes/TssSession.class.ps1]: in 7614719

**Comment by peetrike** [comment]
There are 3 tests that fail for me in New-TssSession tests.  All of them are in old tests and are caused by:
https://github.com/thycotic-ps/thycotic.secretserver/blob/c2a27648ff15ccb088d23a2cb8b6ce558bd9eb8c/tests/functions/New-TssSession.Tests.ps1#L61

https://github.com/thycotic-ps/thycotic.secretserver/blob/c2a27648ff15ccb088d23a2cb8b6ce558bd9eb8c/tests/functions/New-TssSession.Tests.ps1#L65
and
https://github.com/thycotic-ps/thycotic.secretserver/blob/c2a27648ff15ccb088d23a2cb8b6ce558bd9eb8c/tests/functions/New-TssSession.Tests.ps1#L101

In class, the `[datetime]::Now` is used, but in tests `[datetime]::UtcNow` is used instead.  Should these be corrected too?

**Comment by wsmelton** [comment]
Updating the test for that is fine. Did any of the other tests fail as well? Please include the output of the test run or a screenshot.

**Comment by peetrike** [comment]
image
this is only calling New-TssSession.Tests.ps1, with new code

**Comment by peetrike** [comment]
was trying to avoid other tests, as my API user has limited access on our Secret Server instance

**Comment by peetrike** [comment]
> this is only calling New-TssSession.Tests.ps1, with new code

the same as a text:
```
Discovery finished in 24.58s.
[-] New-TssSession works.Credential parameter.Calculates TimeOfDeath 36ms (35ms|1ms)
 Expected the actual value to be less than or equal to 2021-02-12T13:47:05.4536684Z, but got 2021-02-12T16:07:05.1011846+02:00.
 at $sessionCredential.TimeOfDeath | Should -BeLessOrEqual $expectedValue, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:62
 at <ScriptBlock>, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:62
[-] New-TssSession works.Credential parameter.Calculates StartTime 28ms (27ms|1ms)
 Expected the actual value to be less than or equal to 2021-02-12T13:47:05.5294387Z, but got 2021-02-12T15:47:05.1052366+02:00.
 at $sessionCredential.StartTime | Should -BeLessOrEqual $currentTime, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:66
 at <ScriptBlock>, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:66
[-] New-TssSession works.AccessToken parameter.Calculates StartTime 12ms (11ms|1ms)
 Expected the actual value to be less than or equal to 2021-02-12T13:47:07.9306820Z, but got 2021-02-12T15:47:05.1219379+02:00.
 at $sessionAccessToken.StartTime | Should -BeLessOrEqual $currentTime, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:102
 at <ScriptBlock>, C:\Projects\GitHub\thycotic.secretserver\tests\functions\New-TssSession.Tests.ps1:102
Tests completed in 29.43s
Tests Passed: 18, Failed: 3, Skipped: 0 NotRun: 0
```

**Comment by wsmelton** [comment]
Since `New-TssSession` affects the whole module, as every function is dependent upon the output object from the command, I need the whole test framework run to know it will be safe to merge the PR.

If that is not doable I can implement the changes at a later time when I can setup my own lab for this scenario.

**Comment by peetrike** [comment]
I will try full run, bit it might happen, that some tests fail because the user account doesn't have enough privieges...

**Comment by peetrike** [comment]
as i supposed:
```
 48 ❯ Invoke-Pester -Path .\tests\

Starting discovery in 29 files.
WARNING: No Folder found
WARNING: No Folder found
WARNING: Issue creating folder []
System.Management.Automation.RuntimeException: {
  "errorCode": "API_AccessDenied",
  "exceptionType": "API_AccessDenied",
  "stackTrace": "",
  "exceptionMessage": "API_AccessDenied-Access Denied",
  "message": "Access Denied",
  "severity": 0
}
at <ScriptBlock><Process>, C:\Projects\GitHub\thycotic.secretserver\src\parts\ErrorHandling.ps1: line 9
at New-Folder<Process>, C:\Projects\GitHub\thycotic.secretserver\src\functions\folders\New-Folder.ps1: line 109
at <ScriptBlock>, C:\Projects\GitHub\thycotic.secretserver\tests\folders\New-TssFolder.Tests.ps1: line 52
```
I will try to add permissions and gradually fix the expected layout in our server instance...

**Comment by wsmelton** [comment]
There is a [wiki](https://github.com/thycotic-ps/thycotic.secretserver/wiki/Test-Setup) page you can review for the required setup. If there is information missing or not understood you can open a discussion (contributor category) to work through it. (_I want to try and keep PR discussions to the PR change itself_)

**Comment by wsmelton** [comment]
Closing this PR as the feature has been added and released in 0.30.0.

**Pull request closed without merging**